woocommerce/woocommerce-gutenberg-products-block
# Add migration strategy and load blockified templates

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/6513

The goal of this PR is to render the new blockied Product Grid block (when it will be ready), only when the user:
- installs the plugin.
- enables a block theme.

**and don't have a customized template.**

In this table, you can see the expected behavior for the different cases:


Plugin | Theme Type | Blockified block
-- | -- | --
New Installation | Classic Theme | 🚫
New Installation | Block | ✔️
New Installation | Switch from Classic Theme to Block Theme | ✔️
Update | Classic Theme | 🚫
Update | Block | 🚫
Update | Switch from Classic Theme to Block Theme | ✔️


To reach this goal, we created a new class `Migration` that [set the option `wc_blocks_use_blockified_product_grid_block_as_template` to no](https://github.com/woocommerce/woocommerce-blocks/blob/7b16ec17ee79b99bc23f1ff46359d8ffa8875ebf/src/Migration.php/#L49-L54) (this code will be enabled when the new block will be released).

In the class `BlockTemplatesController` [we render the new template only if the option `wc_blocks_use_blockified_product_grid_block_as_template` is set to yes](https://github.com/woocommerce/woocommerce-blocks/blob/7b16ec17ee79b99bc23f1ff46359d8ffa8875ebf/src/BlockTemplatesController.php/#L302-L306).

We added unit tests for the `Migration` class and for the `BlockTemplatesUtils` class.

### Testing

#### Automated Tests

#### User Facing Testing




#### Download zips:

1. Zip generated from the current trunk branch ([link](https://anonfiles.com/VcD5e3v9y0/woocommerce-gutenberg-products-block_zip))
2. Zip generated from the PR. ([link](https://anonfiles.com/96H5e1v8y3/woocommerce-gutenberg-products-block_zip))

**Please, don't rename the name of the zip.**

For convenience, in the testing instructions:
- we use the name `pre-update package` for the zip generated from the current trunk branch.
- we use the name `after-update package` for the zip generated from the PR.



### Test 1: 
Plugin | Theme Type | Blockified block
-- | -- | --
New Installation | Block | ✔️
1. Be sure that you have a block theme installed.
2. Install the plugin with the `after-update package`.
3. Open the FSE -> templates and check that  `Product Catalog`, `Product Search Results`, `Products by Category` and `Products by Tag` renders a `Paragraph block` (it is just a placeholder).

### Test 2: 
Plugin | Theme Type | Blockified block
-- | -- | --
New Installation | Switch from Classic Theme to Block Theme  | ✔️
1. Be sure that you have a classic theme installed.
2. Install the plugin with the `after-update package`.
3. Active a block theme.
4. Open the FSE -> templates and check that  `Product Catalog`, `Product Search Results`, `Products by Category` and `Products by Tag` renders a `Paragraph block` (it is just a placeholder).

### Test 3: 
Plugin | Theme Type | Blockified block
-- | -- | --
Update | Block | 🚫

1. Install `pre-update package`.
2. Open the FSE -> templates and check that  `Product Catalog`, `Product Search Results`, `Products by Category` and `Products by Tag` renders the classic `WooCommerce Product Grid Block`.
3. Update the plugin with `after-update package`.
4. Open the database, and check if in the table `wp_option` there is a key `wc_blocks_use_blockified_product_grid_block_as_template` set to `no`.
5. Open the FSE -> templates and check that  `Product Catalog`, `Product Search Results`, `Products by Category` and `Products by Tag` renders the classic `WooCommerce Product Grid Block`.
6. Remove the key `wc_blocks_use_blockified_product_grid_block_as_template` and the key `wc_blocks_version` from the database and remove the plugin.

### Test 4: 
Plugin | Theme Type | Blockified block
-- | -- | --
Update | Switch from Classic Theme to Block Theme | ✔️

1. Be sure that you have a classic theme installed.
2. Install `pre-update package`.
3. Update the plugin with `after-update package`.
4. Enable a block theme.
5. Open the database, and check if in the table `wp_option` there is a key `wc_blocks_use_blockified_product_grid_block_as_template` set to `yes`.
6. Open the FSE -> templates and check that  `Product Catalog`, `Product Search Results`, `Products by Category`, and `Products by Tag` renders a `Paragraph block` (it is just a placeholder).
7. Remove the key `wc_blocks_use_blockified_product_grid_block_as_template` and the key `wc_blocks_version` from the database and remove the plugin.



### Unit Test Fix
We fixed #6618. We noticed that the command `pretest:php` was executed in another container and break some references in the `phpunit` container. We decide to remove the command: now, before running `npm run php:test` the user must install PHP deps.

### Test: 

Be sure that PHP unit tests don't fail.




### WooCommerce Visibility
